File: kotlin-null-writable/karakum/plugins/convertErrorType.js
import ts from "typescript";

export default function (node) {
    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "Error"
    ) {
        return "js.errors.JsError"
    }

    return null
}

File: kotlin-null-writable/karakum/src/jsMain/kotlin/nullwritable/karakum/plugins/convertErrorType.kt
package nullwritable.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertErrorType = createPlugin { node, _, _ ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Error")

        "js.errors.JsError"
    }
}
